cloudinit/config/schema.py

```diff
@@ -109,6 +157,106 @@ def is_schema_byte_string(checker, instance):
+    if description:
+        msg += f" {description}"
+    return msg
+
+
+def _validator_deprecated(
+    _validator,
+    deprecated: bool,
+    _instance,
+    schema: dict,
+    error_type: Type[Exception],
+):
+    """Jsonschema validator for `deprecated` items.
+
+    It raises a instance of `error_type` if deprecated that must be handled,
+    otherwise the instance is consider faulty.
+    """
+    if deprecated:
+        description = schema.get("description")
+        msg = _add_deprecation_msg(description)
+        yield error_type(msg)
+
+
+def _anyOf(validator, anyOf, instance, _schema, error_type: Type[Exception]):
+    """Jsonschema validator for `anyOf`.
+
+    It treats occurrences of `error_type` as non-errors, but yield them for
+    external processing. Useful to process schema annotations, as `deprecated`.
+    """
+    from jsonschema import ValidationError
+
+    all_errors = []
+    all_deprecations = []
+    for index, subschema in enumerate(anyOf):
+        all_errs = list(
+            validator.descend(instance, subschema, schema_path=index)
+        )
+        errs = list(filter(lambda e: not isinstance(e, error_type), all_errs))
+        deprecations = list(
+            filter(lambda e: isinstance(e, error_type), all_errs)
+        )
+        if not errs:
+            all_deprecations.extend(deprecations)
+            break
+        all_errors.extend(errs)
+    else:
+        yield ValidationError(
+            "%r is not valid under any of the given schemas" % (instance,),
+            context=all_errors,
+        )
+    yield from all_deprecations
+
+
+def _oneOf(validator, oneOf, instance, _schema, error_type: Type[Exception]):
+    """Jsonschema validator for `oneOf`.
+
+    It treats occurrences of `error_type` as non-errors, but yield them for
+    external processing. Useful to process schema annotations, as `deprecated`.
+    """
+    from jsonschema import ValidationError
+
+    subschemas = enumerate(oneOf)
+    all_errors = []
+    all_deprecations = []
+    for index, subschema in subschemas:
+        all_errs = list(
+            validator.descend(instance, subschema, schema_path=index)
+        )
+        errs = list(filter(lambda e: not isinstance(e, error_type), all_errs))
+        deprecations = list(
+            filter(lambda e: isinstance(e, error_type), all_errs)
+        )
+        if not errs:
+            first_valid = subschema
+            all_deprecations.extend(deprecations)
+            break
+        all_errors.extend(errs)
+    else:
+        yield ValidationError(
+            "%r is not valid under any of the given schemas" % (instance,),
+            context=all_errors,
+        )
+
+    more_valid = [s for i, s in subschemas if validator.is_valid(instance, s)]
+    if more_valid:
+        more_valid.append(first_valid)
+        reprs = ", ".join(repr(schema) for schema in more_valid)
+        yield ValidationError(
+            "%r is valid under each of %s" % (instance, reprs)
```

Comment:
Is this correct? It seems odd that this yields a ValidationError that says "this is valid".

Reply:
Yes, it is. When one declares a `oneOf` sub-schema, an instance is valid if and only if it is valid under only one `oneOf`'s sub-schema.

More info: https://json-schema.org/understanding-json-schema/reference/combining.html?highlight=oneof#schema-composition